README.md
# Voice Quality Collector daemon

VQ Collector is simple pure Python daemon that receives SIP Voice Quality reports sent by VoIP devices.
This collector uses the PUBLISH method defined by [sec. 3.2 of RFC 6035](http://tools.ietf.org/html/rfc6035#section-3.2)

## Behavior:
VQ-Collector daemon handles SIP PUBLISH requests containing a body and declaring an *application/vq-rtcpxr* Content-Type.
The PUBLISH body is logged on a local file or trough a remote syslog server

## Supported enviroments:
VQ-Collector is tested on Debian GNU/Linux and OSX 10.7

## Installation:
* you need to configure the installation destination editing the *CONFIG* file
* **DESTDIR** defines the installation directory
* **PYTHON_LIB** defines the directory where all needed python modules will be installed
* **INITDIR** defines the directory where the init script will be placed
* **TARGET** defines the distributrion where the daemon is installed, supported values: *rh* and *deb*, this affects the init script

After editing the CONFIG file you must run **make vq-collector**
Now you're ready to configure the collector

## Configuration:
The configuration file contains some directives under the **[main]** section:

* **local_ip**: this parameter defines where the damon must bind, here you can insert a local IP address, a network interface name or *default*, using *default* the daemon will use the default ip address
* **pid_file**: where the PID will be stored
* **port**: local UDP port to use
* **log_file**: here you can insert a file path or a syslog definition following this syntax: *syslog:IP_ADDRESS:PORT:FACILITY* 
   
   Eg.: *syslog:172.16.18.99:514:local7* will send all messages to the remote syslog on 172.16.18.99, using the 514 UDP port and the local7 facility.
* **reply_to_socket**: setting this paramether to True all PUBLISH reply will be sent to the remote IP/port, otherwise IP and port will be extracted from the PUBLISH *Contact* header 
* **debug**: here you can insert *True* or *False*, use *True* only during debug purpose
* **daemon**: another boolean value, using *False* the program will start in foreground

## Usage:
You can run the collector in this way:
 
    ./vq-collector -s /etc/vq-collector.conf
    
Or using the init scritp

    /etc/init.d/vq-collector start

**NOTE:** the init script is very simple and need some adjustements in order to run in your distribution.

# VQ-Session report syntax:

Reports syntax is defined by [RFC3611](http://tools.ietf.org/html/rfc3611)


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/pbertera/vq-collector/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



changelog
vq-collector (0.0.1-2) stable; urgency=low

 * added RH script
 
 -- Pietro Bertera <pietro@bertera.it>  Mon, 6 Sep 2013 15:24:11 +0100

vq-collector (0.0.1-1) stable; urgency=low

  * Initial packaging

 -- Pietro Bertera <pietro@bertera.it>  Mon, 6 Sep 2013 15:24:11 +0100


src/siplib/sip.py
#!/usr/bin/python
# vi:si:et:sw=4:sts=4:ts=4
# -*- coding: UTF-8 -*-
# -*- Mode: Python -*-
#
# Copyright (c) 2013 Pietro Bertera <pietro@bertera.it>
# This file is covered by the LGPLv3 or later, read COPYING for details.

import cStringIO


class SipError(Exception):
    pass


class SipUnpackError(SipError):
    pass


class SipNeedData(SipUnpackError):
    pass


class SipPackError(SipError):
    pass


def canon_header(s):
    exception = {'call-id': 'Call-ID', 'cseq':
                 'CSeq', 'www-authenticate': 'WWW-Authenticate'}
    short = ['allow-events', 'u', 'call-id', 'i', 'contact', 'm',
             'content-encoding', 'e', 'content-length', 'l',
             'content-type', 'c', 'event', 'o', 'from', 'f',
             'subject', 's', 'supported', 'k', 'to', 't', 'via', 'v']
    s = s.lower()
    return ((len(s) == 1) and s in short and canon_header(short[short.index(s) - 1])) \
        or (s in exception and exception[s]) or '-'.join([x.capitalize() for x in s.split('-')])


def parse_headers(f):
    """Return dict of HTTP headers parsed from a file object."""
    d = {}
    while 1:
        line = f.readline()
        line = line.strip()
        if not line:
            break
        l = line.split(None, 1)
        if not l[0].endswith(':'):
            raise SipUnpackError('invalid header: %r' % line)
        k = l[0][:-1].lower()
        d[k] = len(l) != 1 and l[1] or ''
    return d


def parse_body(f, headers):
    """Return SIP body parsed from a file object, given HTTP header dict."""
    if 'content-length' in headers:
        n = int(headers['content-length'])
        body = f.read(n)
        if len(body) != n:
            raise SipNeedData(
                'short body (missing %d bytes)' % (n - len(body)))
    elif 'content-type' in headers:
        body = f.read()
    else:
        body = ''
    return body


class Message:

    """SIP Protocol headers + body."""
    __metaclass__ = type
    __hdr_defaults__ = {}
    headers = None
    body = None

    def __init__(self, *args, **kwargs):
        if args:
            self.unpack(args[0])
        else:
            self.headers = {}
            self.body = ''
            for k, v in self.__hdr_defaults__.iteritems():
                setattr(self, k, v)
            for k, v in kwargs.iteritems():
                setattr(self, k, v)

    def unpack(self, buf):
        f = cStringIO.StringIO(buf)
        # Parse headers
        self.headers = parse_headers(f)
        # Parse body
        self.body = parse_body(f, self.headers)
                # Save the rest
        self.data = f.read()

    def pack_hdr(self):
        return ''.join(['%s: %s\r\n' % (canon_header(k), v) for k, v in self.headers.iteritems()])

    def __len__(self):
        return len(str(self))

    def __str__(self):
        return '%s\r\n%s' % (self.pack_hdr(), self.body)


class Request(Message):

    """SIP request."""
    __hdr_defaults__ = {
        'method': 'INVITE',
        'uri': 'sip:user@example.com',
        'version': '2.0',
        'headers': {'to': '', 'from': '', 'call-id': '', 'cseq': '', 'contact': ''}
    }
    __methods = dict.fromkeys((
                              'ACK', 'BYE', 'CANCEL', 'INFO', 'INVITE', 'MESSAGE', 'NOTIFY',
                              'OPTIONS', 'PRACK', 'PUBLISH', 'REFER', 'REGISTER', 'SUBSCRIBE',
                              'UPDATE'
                              ))
    __proto = 'SIP'

    def unpack(self, buf):
        f = cStringIO.StringIO(buf)
        line = f.readline()
        l = line.strip().split()
        if len(l) != 3 or l[0] not in self.__methods or \
                not l[2].startswith(self.__proto):
            raise SipUnpackError('invalid request: %r' % line)
        self.method = l[0]
        self.uri = l[1]
        self.version = l[2][len(self.__proto) + 1:]
        Message.unpack(self, f.read())

    def __str__(self):
        return '%s %s %s/%s\r\n' % (self.method, self.uri, self.__proto,
                                    self.version) + Message.__str__(self)


class Response(Message):

    """SIP response."""
    __hdr_defaults__ = {
        'version': '2.0',
        'status': '200',
        'reason': 'OK',
        'headers': {'to': '', 'from': '', 'call-id': '', 'cseq': '', 'contact': ''}
    }
    __proto = 'SIP'

    def unpack(self, buf):
        f = cStringIO.StringIO(buf)
        line = f.readline()
        l = line.strip().split(None, 2)
        if len(l) < 2 or not l[0].startswith(self.__proto) or not l[1].isdigit():
            raise SipUnpackError('invalid response: %r' % line)
        self.version = l[0][len(self.__proto) + 1:]
        self.status = l[1]
        self.reason = l[2]
        Message.unpack(self, f.read())

    def __str__(self):
        return '%s/%s %s %s\r\n' % (self.__proto, self.version, self.status,
                                    self.reason) + Message.__str__(self)


src/vq-collector/vq-collector.py
#!/usr/bin/python
# vi:si:et:sw=4:sts=4:ts=4
# -*- coding: UTF-8 -*-
# -*- Mode: Python -*-
#
# Copyright (c) 2013 Pietro Bertera <pietro@bertera.it>
# This file is covered by the LGPLv3 or later, read COPYING for details.

import socket
import struct
import re
import sys
import select
import ConfigParser
import os
import signal
import cStringIO
import importlib
from string import Template

from siplib import sip
from siplib import daemon

version = "__VERSION__"
logger = daemon.Logger(sys.argv[0])


class SocketOptionError(Exception):
    pass


class CreateSocketError(Exception):
    pass


class BindSocketError(Exception):
    pass


class SendDataError(Exception):
    pass


class CollectorServer:

    def __init__(self, local_ip, port=5060, reply_to_socket=False, feedback_callback=False, feedback_module_dir=None):
        self.port = port
        self.local_ip = local_ip
        self.listen_addr = local_ip
        self.reply_to_socket = reply_to_socket
        self.feedback_callback = feedback_callback
        self.feedback_module_dir = feedback_module_dir

        logger.debugMessage("Local IP: %s" % self.local_ip)
        logger.debugMessage("Listen IP: %s" % self.listen_addr)

        self.recvsocket = self._create_socket()
        
        if self.feedback_callback:
            if self.feedback_module_dir:
                sys.path.append(os.path.abspath(self.feedback_module_dir))
            from vq_callback import vq_callback
            self.vq_callback = vq_callback

    def listen(self):
        # Sockets from which we expect to read
        inputs = [self.recvsocket]
        # Sockets to which we expect to write
        outputs = []

        logger.debugMessage("Starting listening loop")

        while inputs:
            readable, writable, exceptional = select.select(inputs, outputs,
                                                            inputs)
            # Handle inputs
            for s in readable:
                if s is self.recvsocket:
                    if not self.handle_request():
                        continue

    def handle_request(self):
        data, remote = self.recvsocket.recvfrom(10240)
        try:
            request = sip.Request(data)
        except sip.SipUnpackError:
            return False

        logger.debugMessage("Received request from %s:%d : \n%s" %
                            (remote[0], remote[1], str(request)))
        response = sip.Response()
        # logging the body
        logger.infoMessage(request.body.split())
        if request.method != "PUBLISH":
            logger.debugMessage("Received a non PUBLISH: %s" %
                                request.method)
            response.reason = "Not implemented"
            response.status = "501"
        else:
            try:
                if request.headers["content-type"] != "application/vq-rtcpxr":
                    logger.debugMessage("No VQ RTCP-XR body detected")
                    response.reason = "Not implementented"
                    response.status = "501"

            except KeyError:
                logger.debugMessage("No VQ RTCP-XR body detected")
                response.reason = "Not implementented"
                response.status = "501"

        response.headers['from'] = request.headers['from']
        response.headers['to'] = request.headers['to']
        response.headers['call-id'] = request.headers['call-id']
        response.headers['cseq'] = request.headers['cseq']
        response.headers['expires'] = 0
        response.headers['via'] = request.headers['via']
        response.headers['contact'] = "<sip:%s:%d;transport=tcp;handler=dum>"\
                                      % (self.local_ip, self.port)
        response.headers['content-length'] = 0
        # Regexp parsing via Header: SIP/2.0/UDP 172.16.18.90:5060;rport
        p = re.compile(r'SIP/(.*)/(.*)\s(.*):([0-9]*);*')
        m = p.search(request.headers['via'])
        
        if m:
            version = m.group(1)
            transport = m.group(2)
            if version != "2.0":
                UnsupportedSIPVersion(
                    "Unsupported SIP version in Via header: %s" % version)
                return false
        else:
            sendDataError("Wrong Via: header")
            return False
        
        if self.reply_to_socket == False:
            p = re.compile(r'<sip:(.*)@(.*):([0-9]+);*(.*)>')
            m = p.search(request.headers['contact'])
            if m:
                phone_ip = m.group(2)
                phone_port = m.group(3)
                logger.debugMessage("Phone IP and port from Contact header: %s:%s" %(phone_ip, phone_port))
        else:
            phone_ip = remote[0]
            phone_port = remote[1]
            logger.debugMessage("Phone IP and port from socket: %s:%d" %(phone_ip, phone_port))
        
        if transport.upper() == "UDP":
            self.send_response(phone_ip, phone_port, response)
        else:
            UnsupportedSIPTransport("Unsupported Transport in Via: header")
            return False
        if self.feedback_callback:
            logger.infoMessage("Executing the callback vqcallback()")
            self.vq_callback((phone_ip, phone_port), request.body)
    
    def send_response(self, phone_ip, phone_port, response):
        logger.debugMessage("Creating send socket")
        try:
            self.sendsock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM,
                                          socket.IPPROTO_UDP)
        except Exception, e:
            CreateSocketError("Cannot create socket: %s" % e)
        try:
            self.sendsock.setsockopt(socket.SOL_SOCKET,
                                     socket.SO_REUSEADDR, 1)
            self.sendsock.setsockopt(socket.SOL_SOCKET,
                                     socket.SO_REUSEPORT, 1)
        except AttributeError, e:
            pass
        try:
            logger.debugMessage("Binding to local ip:port %s:%s" %
                                (self.local_ip, self.port))
            self.sendsock.bind((self.local_ip, self.port))
        except Exception, e:
            SendDataError("Cannot bind socket to %s:%d: %s"
                          % (self.local_ip, self.port, e))

        # sent the OK (or 501)
        try:
            logger.debugMessage("Sending response to %s:%s : \n%s" %
                                (phone_ip, phone_port, str(response)))
            self.sendsock.sendto(str(response), (phone_ip,
                                                 int(phone_port)))
        except Exception, e:
            SendDataError("Cannot send OK/DENY response to %s:%s: %s" %
                          (phone_ip, phone_port, e))
        self.sendsock.close()

    def _create_socket(self):
        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM,
                                 socket.IPPROTO_UDP)
            sock.setblocking(0)
        except Exception, e:
            raise CreateSocketError("Cannot create socket: %s" % e)
        try:
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        except AttributeError:
            pass
        try:
            sock.bind((socket.gethostbyname(self.listen_addr), self.port))
        except Exception, e:
            raise BindSocketError("Cannot bind socket to %s:%d: %s" %
                                  (self.listen_addr, self.port, e))
        return sock

if __name__ == '__main__':

    def usage():
        print "\nUsage: %s [options] <config-file>" % sys.argv[0]
        print "\n\tOptions:"
        print "\t\t-s <config-file>\t\tStart the program"
        sys.exit()

    def signal_handler(signal, frame):
        logger.infoMessage('Killed by SIGTERM. Goodbye.')
        sys.exit(0)

    signal.signal(signal.SIGTERM, signal_handler)

    # TODO: improve command line handling
    if len(sys.argv) < 3:
        print "\nERROR: missing args."
        usage()

    if sys.argv[1] not in ["-c", "-s"]:
        print "\nERROR: wrong arg."
        usage()

    parser = daemon.SettingsParser(sys.argv[2], logger)
    config = parser.get_config()
    main_settings = parser.parse_main()

    if main_settings["debug"].upper() == 'TRUE':
        logger.set_level("debug")
        logger.debugMessage("Log level: DEBUG")
    else:
        logger.set_level("info")
        logger.debugMessage("Log level: INFO")

    if main_settings["reply_to_socket"].upper() == 'TRUE':
        logger.debugMessage("All replies will be sent over socket")
        main_settings["reply_to_socket"] = True
    else:
        main_settings["reply_to_socket"] = False

    local_ip = main_settings["local_ip"]

    if main_settings["conf_log_handler"]:
        logger.debugMessage("Logging to %s" % main_settings["log_file"])
        logger.change_handler(main_settings["conf_log_handler"])
    # non-common setting
    try:
        port = int(config.get('main', 'port'))
        logger.infoMessage("Listening to %d port" % int(port))
    except ConfigParser.NoOptionError:
        port = 5060
    try:
        feedback_callback = config.get('main', 'feedback_callback')
        if feedback_callback.upper() == 'TRUE':
            feedback_callback = True
        else:
            feedback_callback = False
        logger.infoMessage("Feedback callback: %s" % feedback_callback)
    except ConfigParser.NoOptionError:
        feedback_callback = False
    try:
        feedback_module_dir = config.get('main', 'feedback_module_dir')
        logger.infoMessage("Feedback module dir: %s" % feedback_module_dir)
    except ConfigParser.NoOptionError:
        feedback_module_dir = None
    

    server = CollectorServer(\
                local_ip=local_ip, port=port,\
                reply_to_socket=main_settings["reply_to_socket"],\
                feedback_callback=feedback_callback,\
                feedback_module_dir=feedback_module_dir\
                )

    if sys.argv[1] == "-s":
        try:
            if config.get('main', 'daemon').upper() == 'TRUE':
                logger.infoMessage('Daemonizing')
                try:
                    pid_file = config.get('main', 'pid_file')
                    logger.infoMessage('Using pid file %s' % pid_file)
                    try:
                        pid = daemon.become_daemon(pid_file)
                    except Exception, e:
                        logger.systemError("Cannot start daemon: %s, exiting"
                                           % e)
                        sys.exit(-1)
                except ConfigParser.NoOptionError:
                    try:
                        pid = daemon.become_daemon(None)
                    except Exception, e:
                        logger.systemError("Cannot start daemon: %s, \
                                            exiting" % e)
                        sys.exit(-1)
                    logger.infoMessage("No pid file in configuration \
                                               file")
                logger.infoMessage("Daemon started with pid %d"
                                   % pid)

        except ConfigParser.NoOptionError:
            logger.infoMessage('Run in foreground')
        try:
            server.listen()
        except Exception, e:
            logger.systemError('Received fatal exception: %s' % e)


src/siplib/daemon.py
#!/usr/bin/python
# vi:si:et:sw=4:sts=4:ts=4
# -*- coding: UTF-8 -*-
# -*- Mode: Python -*-
#
# Copyright (c) 2013 Pietro Bertera <pietro@bertera.it>
# This file is covered by the LGPLv3 or later, read COPYING for details.

import struct
import socket
import logging
import logging.handlers
import sys
import os
import re
import ConfigParser

from string import Template


class InvalidRegexError(Exception):
    pass


class Logger:

    def __init__(self, name, level="info"):
        logging.handlers.raiseExceptions = True

        self.logger = logging.getLogger(name)
        self.log_handler = logging.StreamHandler()
        self.formatter = logging.Formatter(
            '%(asctime)s %(levelname)s %(message)s')
        self.log_handler.setFormatter(self.formatter)
        self.logger.addHandler(self.log_handler)
        if level == "info":
            self.logger.setLevel(logging.INFO)
        if level == "debug":
            self.logger.setLevel(logging.DEBUG)

    def change_handler(self, new_handler):
        new_handler.setFormatter(self.formatter)
        self.logger.removeHandler(self.log_handler)
        self.logger.addHandler(new_handler)

    def set_level(self, level):
        if level == "info":
            self.logger.setLevel(logging.INFO)
        if level == "debug":
            self.logger.setLevel(logging.DEBUG)

    def error(self, message=None):
        if message:
            self.logger.error(message)
        pass

    unsupportedSIPVersion = error
    unsupportedSIPTransport = error
    sendDataError = error
    systemError = error
    configError = error

    def debug(self, message=None):
        if message:
            self.logger.debug(message)
        pass

    debugMessage = debug

    def info(self, message=None):
        if message:
            self.logger.info(message)
        pass

    infoMessage = info

# errors


class SocketOptionError(Exception):
    pass


class CreateSocketError(Exception):
    pass


class BindSocketError(Exception):
    pass


class InvalidRegexError(Exception):
    pass


class SettingsParser:

    def __init__(self, config_file, logger=None):

        if not logger:
            self.logger = Logger("SettingsParser")

        self.logger = logger
        self.config_file = config_file
        self.config = ConfigParser.RawConfigParser()
        # Make case sensitive parser
        self.config.optionxform = str
        self.config.read(self.config_file)

        self.common_config = {}

    def get_config(self):
        return self.config

    def parse_main(self):
        """
        This function parse the configuration file for
        - debug: "True", "False"
        - local_ip: "default", "xxx.xxx.xxx.xxx", "eth0"
        - log_file: "/var/log/file.log", "syslog://"
        - daemon: "True", "False"
        - pid_file: "/var/run/test.pid"
        """
        try:
            self.common_config["debug"] = self.config.get('main', 'debug')
        except ConfigParser.NoOptionError:
            self.common_config["debug"] = "FALSE"

        try:
            conf_local_ip = self.config.get('main', 'local_ip')
            if is_valid_ipv4_address(conf_local_ip):
                self.common_config["local_ip"] = conf_local_ip

            #if loca_if == "default" try to reach google.com
            elif conf_local_ip == "default":
                try:
                    self.common_config["local_ip"] = get_ip_address()
                except Exception, e:
                    self.logger.configError(
                        "cannot discover local ip address: %s" % e)
                    sys.exit(1)

            else:  # network interface name
                try:
                    self.common_config[
                        "local_ip"] = get_ip_address_ifname(conf_local_ip)
                except Exception, e:
                    self.logger.configError(
                        "cannot determine ip address of %s interface: %s" % (conf_local_ip, e))
                    sys.exit(1)

        except ConfigParser.NoOptionError:
            self.logger.configError(
                "Missing mandatory parameters in config file, bailing out!")
            sys.exit(1)

        try:
            log_file = self.common_config[
                "log_file"] = self.config.get('main', 'log_file')
            if log_file.startswith("syslog"):
                try:
                    syslog_host = log_file.split(":")[1]
                except IndexError:
                    syslog_host = 'localhost'
                try:
                    syslog_port = int(log_file.split(":")[2])
                except IndexError:
                    syslog_port = 514
                try:
                    syslog_facility = log_file.split(":")[3]
                except IndexError:
                    syslog_facility = logging.handlers.SysLogHandler.LOG_USER
                self.logger.debugMessage(
                    "Logging to syslog (host: %s, port: %s, facility: %s)" %
                    ((syslog_host, syslog_port, syslog_facility)))
                self.common_config["conf_log_handler"] = logging.handlers.SysLogHandler(
                    (syslog_host, syslog_port), syslog_facility)
            else:
                self.logger.debugMessage("Logging to file: %s" % log_file)
                try:
                    self.common_config[
                        "conf_log_handler"] = logging.FileHandler(log_file)
                except IOError, e:
                    self.logger.configError(
                        "cannot access to the log file: %s" % e)
                    sys.exit(1)

        except ConfigParser.NoOptionError:
            # no log defined in config file
            self.common_config["conf_log_handler"] = None

        try:
            self.common_config["daemon"] = self.config.get('main', 'daemon')
        except ConfigParser.NoOptionError:
            self.common_config["daemon"] = None
        try:
            self.common_config[
                "pid_file"] = self.config.get('main', 'pid_file')
        except ConfigParser.NoOptionError:
            self.common_config["pid_file"] = None
        
        try:
            self.common_config["reply_to_socket"] = self.config.get('main', 'reply_to_socket')
        except ConfigParser.NoOptionError:
            self.common_config["reply_to_socket"] = None

        return self.common_config

# daemonizing function


def become_daemon(pid_file=None):
    try:
        pid = os.fork()
        if pid > 0:
            sys.exit(0)
    except OSError, e:
        raise Exception("Fork failed, can't become daemon: %d (%s)" %
                        (e.errno, e.strerror))
        sys.exit(1)
    os.chdir("/")
    os.setsid()
    os.umask(0)

    pid = os.getpid()
    if pid_file:
        try:
            f = open(pid_file, "w")
            f.write(str(pid) + "\n")
            f.close()
        except IOError:
            raise Exception('Cannot write PID in pidfile %s' % pid_file)
    return pid


# net functions
def get_ip_address():
    # This is a simple hack to find our IP address
    # AFAIK this is the only platform-independent way to obtain the address
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(('google.com', 80))
    ip = s.getsockname()[0]
    s.close()
    return ip


def is_valid_ipv4_address(address):
    try:
        addr = socket.inet_pton(socket.AF_INET, address)
    except AttributeError:  # no inet_pton here, sorry
        try:
            addr = socket.inet_aton(address)
        except socket.error:
            return False
        #invalidate shortened address (like 127.1)
        return address.count('.') == 3
    except socket.error:  # not a valid address
        return False
    return True


def get_ip_address_ifname(iface):
    import fcntl
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    return socket.inet_ntoa(fcntl.ioctl(
        s.fileno(),
        0x8915,  # SIOCGIFADDR
        struct.pack('256s', iface[:15])
    )[20:24])
